### ISSUE TITLE
Array map changes previously inferred/declared key type

### ISSUE THREAD

**Issue by sserbin** [vimeo/psalm#541]
Test case:
```php
/**
 * @psalm-return array{key1:int,key2:int}
 */
function foo(): array {
  $v = ["key1"=> 1, "key2"=> "2"];
  $r = array_map('intval', $v);
  return $r;
}
```

Expected: no errors
Actual:
```
The declared return type 'array{key1:int, key2:int}' for foo is more specific than the inferred return type 'array<int, int>'
```

As stated in `array_map` doc:
> The returned array will preserve the keys of the array argument if and only if exactly one array is passed. If more than one array is passed, the returned array will have sequential integer keys.

**Comment by muglug** [comment]
Thanks!
